### ISSUE TITLE
cmake compilation: find vorbis error

### ISSUE THREAD

**Issue by mipac** [erikd/libsndfile#530]
When using cmake to compile the lib with externals lib,
I can't make working find vorbis
my vorbis, vorbisenc, and vorbisfile are compiled and installed in custom dir (say ~/target)

I try to set PKG_CONFIG_PATH, CMAKE_PREFIX_PATH to ~/target

it works with ogg, opus and flac but vorbis is not found...

trying with configure is ok with --prefix

regards

**Comment by evpobr** [comment]
Hello, @mipac, i need more info:

1. Target OS.
2. Versions of libsndfile and  external libs.
3. How external libs were installed (CMake or Autotools).
4. CMake configuration log.
5. Build logs.

**Comment by mipac** [comment]
Hello,

On Debian 10, with a fresh git version of sndfile and xiph external libs
oog, opus, flac intalled with their cmake installs in the same target dir
I see that the findXXX.cmake where based on pkg-config, so I've try to set PKG_CONFIG_PATH to the .pc files

but anyway, sndfile cmake call can't find vorbis and exlude external libs from compilation

...
-- Found ALSA: /usr/lib/x86_64-linux-gnu/libasound.so (found version "1.1.8") 
-- Found Sndio: /usr/lib/x86_64-linux-gnu/libsndio.so  
-- Found Ogg: /home/XXX/DEV/targets/amd64/lib/libogg.so (found suitable version "1.3.4", minimum required is "1.3") 
-- Could NOT find Vorbis (missing: Enc) 
-- Found FLAC: /home/XXX/DEV/targets/amd64/lib/libFLAC.so  
-- Found Opus: /home/XXX/DEV/targets/amd64/lib/libopus.so  
...

the pc files of xiph libs are similar, I don't understand why vorbis is not found...
I try to delete the QUIET flag in the findVorbis.cmake but not infos...

with autotools it's ok, using the good --prefix

**Comment by evpobr** [comment]
Can you check with `CMAKE_FIND_PACKAGE_PREFER_CONFIG` set to `ON`? Don't forget to delete build directory first.

**Comment by evpobr** [comment]
Cannot reproduce your issue on Windows platform. This is my configure output:

```
[cmake] -- Found Ogg: C:/Users/evpob/source/repos/vcpkg/installed/x64-windows-static/lib/ogg.lib (Required is at least version "1.3") 
[cmake] -- Found Vorbis: C:/Users/evpob/source/repos/vcpkg/installed/x64-windows-static/lib/vorbis.lib  found components: Enc 
[cmake] -- Found FLAC: C:/Users/evpob/source/repos/vcpkg/installed/x64-windows-static/lib/FLAC.lib  
[cmake] -- Found Opus: C:/Users/evpob/source/repos/vcpkg/installed/x64-windows-static/lib/opus.lib  
[cmake] -- Found Speex: C:/Users/evpob/source/repos/vcpkg/installed/x64-windows-static/lib/libspeex.lib  
```

**Comment by Magnutic** [comment]
Hi, I can reproduce this issue on Windows. I built and installed ogg and vorbis to C:\repos\install and set that as CMAKE_PREFIX_PATH when configuring libsndfile.

CMAKE_FIND_PACKAGE_PREFER_CONFIG=1 indeed works.
I guess there is some difference between the vckpg installation and a manual build.

All libraries were checked out from github last week:
ogg at 0bbcba4
vorbis at tag 1.3.7
libsndfile at 2ccb23f

I attached configure and build/install logs below, as well as the install tree.

[install-tree.txt](https://github.com/erikd/libsndfile/files/4961759/install-tree.txt)
[libsndfile-configure-log.txt](https://github.com/erikd/libsndfile/files/4961760/libsndfile-configure-log.txt)
[libsndfile-configure-prefer-config-log.txt](https://github.com/erikd/libsndfile/files/4961761/libsndfile-configure-prefer-config-log.txt)
[ogg-build-log.txt](https://github.com/erikd/libsndfile/files/4961762/ogg-build-log.txt)
[ogg-configure-log.txt](https://github.com/erikd/libsndfile/files/4961763/ogg-configure-log.txt)
[vorbis-build-log.txt](https://github.com/erikd/libsndfile/files/4961764/vorbis-build-log.txt)
[vorbis-configure-log.txt](https://github.com/erikd/libsndfile/files/4961765/vorbis-configure-log.txt)

**Comment by evpobr** [comment]
Thanks! I need also `CMakeCache.txt` for build without `CMAKE_FIND_PACKAGE_PREFER_CONFIG`, for clean configuration (delete build directory first).

**Comment by Magnutic** [comment]
Sure, attached below:
[CMakeCache.txt](https://github.com/erikd/libsndfile/files/4967362/CMakeCache.txt)

**Comment by ellenjohnson** [comment]
Hello!
  I'm having same problem with libsndfile not finding vorbis.  I'm using CMake on Windows10.
  I've compiled the latest versions of ogg, flac, and vorbis using CMake. 
  I have vorbis, vorbisenc, and vorbisfile compiled and installed.

  For libsndfile CMake options to find the dependent libraries, I'm setting these options to point to their respective locations:  OGG_INCLUDE_DIR, OGG_LIBRARY, FLAC_INCLUDE_DIR, FLAC_LIBRARY, VORBIS_INCLUDE_DIR, and VORBIS_LIBRARY.  It finds ogg and flac okay, but not vorbis:

-- Could NOT find Vorbis (missing: Vorbis_Vorbis_INCLUDE_DIR Enc)

I see there's a FindVorbis.cmake and FindVorbisEnc.cmake, and looks like FindVorbis.cmake was modified a couple of months ago.  To find vorbis, are the options not VORBIS_INCLUDE_DIR and VORBIS_LIBRARY?  Are there more I need to specify (i.e., for vorbisenc?)
Thanks!

**Comment by evpobr** [comment]
>> For libsndfile CMake options to find the dependent libraries, I'm setting these options to point to their respective locations:  OGG_INCLUDE_DIR, OGG_LIBRARY, FLAC_INCLUDE_DIR, FLAC_LIBRARY, VORBIS_INCLUDE_DIR, and VORBIS_LIBRARY. It finds ogg and flac okay, but not vorbis:

>> -- Could NOT find Vorbis (missing: Vorbis_Vorbis_INCLUDE_DIR Enc)

You need to use `Vorbis_Vorbis_XXX` variables, not `VORBIS_XXX`, or `Vorbis_ROOT`.

**Comment by evpobr** [comment]
Still cannot reproduce bug (Windows 10 + Visual Studio 2019).

Without `CMAKE_FIND_PACKAGE_PREFER_CONFIG`

```
[proc] Выполнение команды: "C:\Program Files\CMake\bin\cmake.EXE" --no-warn-unused-cli -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_INSTALL_PREFIX:STRING=D:/temp/install -Hc:/Users/evpob/source/repos/libsndfile -Bc:/Users/evpob/source/repos/libsndfile/build -G Ninja
[cmake] Not searching for unused variables given on the command line.
[cmake] -- The C compiler identification is MSVC 19.26.28806.0
[cmake] -- The CXX compiler identification is MSVC 19.26.28806.0
[cmake] -- Detecting C compiler ABI info
[cmake] -- Detecting C compiler ABI info - done
[cmake] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.26.28801/bin/Hostx64/x64/cl.exe - skipped
[cmake] -- Detecting C compile features
[cmake] -- Detecting C compile features - done
[cmake] -- Detecting CXX compiler ABI info
[cmake] -- Detecting CXX compiler ABI info - done
[cmake] -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.26.28801/bin/Hostx64/x64/cl.exe - skipped
[cmake] -- Detecting CXX compile features
[cmake] -- Detecting CXX compile features - done
[cmake] -- 
[cmake] -- 
[cmake] -- Checking large files support...
[cmake] -- 
[cmake] -- Result of checking large files support: supported with WinAPI
[cmake] -- Found Ogg: D:/temp/install/lib/ogg.lib (Required is at least version "1.3") 
[cmake] -- Found Vorbis: D:/temp/install/lib/vorbis.lib  found components: Enc 

...

[cmake] -- Configuring done
[cmake] -- Generating done
[cmake] -- Build files have been written to: C:/Users/evpob/source/repos/libsndfile/build
```

With `CMAKE_FIND_PACKAGE_PREFER_CONFIG=TRUE`

```
[proc] Выполнение команды: "C:\Program Files\CMake\bin\cmake.EXE" --no-warn-unused-cli -DCMAKE_FIND_PACKAGE_PREFER_CONFIG:BOOL=TRUE -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_INSTALL_PREFIX:STRING=D:/temp/install -Hc:/Users/evpob/source/repos/libsndfile -Bc:/Users/evpob/source/repos/libsndfile/build -G Ninja
[cmake] Not searching for unused variables given on the command line.
[cmake] -- The C compiler identification is MSVC 19.26.28806.0
[cmake] -- The CXX compiler identification is MSVC 19.26.28806.0
[cmake] -- Detecting C compiler ABI info
[cmake] -- Detecting C compiler ABI info - done
[cmake] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.26.28801/bin/Hostx64/x64/cl.exe - skipped
[cmake] -- Detecting C compile features
[cmake] -- Detecting C compile features - done
[cmake] -- Detecting CXX compiler ABI info
[cmake] -- Detecting CXX compiler ABI info - done
[cmake] -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.26.28801/bin/Hostx64/x64/cl.exe - skipped
[cmake] -- Detecting CXX compile features
[cmake] -- Detecting CXX compile features - done
[cmake] -- 
[cmake] -- 
[cmake] -- Checking large files support...
[cmake] -- 
[cmake] -- Result of checking large files support: supported with WinAPI
[cmake] -- Could NOT find FLAC (missing: FLAC_DIR)
[cmake] -- Could NOT find Opus (missing: Opus_DIR)
[cmake] -- Found Ogg: D:/temp/install/lib/cmake/Ogg/OggConfig.cmake (found version "1.3.4") 
[cmake] -- Found Vorbis: D:/temp/install/lib/cmake/Vorbis/VorbisConfig.cmake (found version "1.3.7") 

...

[cmake] -- Configuring done
[cmake] -- Generating done
[cmake] -- Build files have been written to: C:/Users/evpob/source/repos/libsndfile/build
```

**Comment by evpobr** [comment]
> Sure, attached below:
> [CMakeCache.txt](https://github.com/erikd/libsndfile/files/4967362/CMakeCache.txt)

Very strange, vorbisenc is found (`CMakeCache.txt`):

```
//Path to a file.
Vorbis_Enc_INCLUDE_DIR:PATH=C:/repos/install/include

//Path to a library.
Vorbis_Enc_LIBRARY:FILEPATH=C:/repos/install/lib/vorbisenc.lib
```

**Comment by evpobr** [comment]
@mipac, did you copy required find modules from `libsndfile/cmake` directory to your project modules directory?

**Comment by mipac** [comment]
no I don't
but I missed to mention is that I use vorbis compiled from their official
git

I've tried to compile sndfile and all dependencies from sources

**Comment by evpobr** [comment]
@mipac, bug confirmed on Debian 10, working on it.

**Comment by evpobr** [comment]
Testing branch is [here](https://github.com/evpobr/libsndfile/commits/fix-cmake-vorbis-handling). @mipac, @Magnutic, @ellenjohnson test please.

**Comment by Magnutic** [comment]
The branch works for me.

But I did some digging and I believe I have found the root cause of the issue:
The FindVorbis.cmake module passes 'NAME_MISMATCHED' when calling find_package_handle_standard_args.
This parameter appears to have been added in CMake 3.17: https://cmake.org/cmake/help/v3.17/module/FindPackageHandleStandardArgs.html

I was using CMake 3.16.4.
I guess the effect was that NAME_MISMATCHED was interpreted as one of the REQUIRED_VARS.
